### PR TITLE
Adds the programs database and ansible template vars

### PR DESCRIPTION
**Pull request by pomegranited**
The `programs` app has been added to `open-release/ficus.master`'s [`edx_sandbox.yml`](https://github.com/edx/configuration/blob/open-release/ficus.master/playbooks/edx_sandbox.yml#L49).  To support this role, we need to add the programs database and user, and set the corresponding ansible variables.

**Testing instructions**:

1. Create a new instance with settings similar to [ficus-xenial AppServer 7](https://console.opencraft.com/instance/2420/edx-appserver/782/):

    ```
    configuration_source_repo_url:	https://github.com/open-craft/configuration 
    configuration_version:	jill/ficus.master
    edx_platform_repository_url:	https://github.com/open-craft/edx-platform
    edx_platform_commit:	jill/ficus.master
    openedx_release:	open-release/ficus.master

    configuration_extra_settings: |	
      # configuration PR ##3522 changes openstack role variables
      COMMON_ENABLE_OPENSTACK_INTEGRATION: true
      COMMON_SWIFT_LOG_SYNC: true

      # Have to specify settings as openstack, not aws
      COMMON_EDXAPP_SETTINGS: 'openstack'
    ```
    Note that AppServer 7 failed to provision due to an error in `TASK [programs : migrate]`.
1. Spawn an appserver.  Provisioning should succeed beyond the `[programs]` role.
1. Spawn a second appserver, and ensure that the 3 "Provisioning" steps runs (again) without error.

**Author notes and concerns**:

1. The `programs` role has not yet been added to [`edx-stateless.yml`](https://github.com/edx/configuration/blob/open-release/ficus.master/playbooks/edx-stateless.yml), but since our clients might want the role, it's worth adding support for it into OCIM.

**Reviewers**
- [ ] @mtyaka

### CODE REVIEW

File: instance/models/mixins/openedx_database.py [hunk 1]
```diff
@@ -95,6 +95,10 @@ def mysql_databases(self):
                     }
                 ],
             },
+            {
+                "name": self._get_mysql_database_name("programs"),
+                "user": self._get_mysql_user_name("program"),
+            },
```
Comment (antoviaque) [inline instance/models/mixins/openedx_database.py]: Would that automatically create the new database for existing instances?

Comment (pomegranited) [inline instance/models/mixins/openedx_database.py]: Ah, good point.  Instances with previously provisioned databases have their [`mysql_provisioned` flag set, and so the databases aren't re-created](https://github.com/open-craft/opencraft/blob/f94660e9e267f85af61ceb50c6600003998b6276/instance/models/mixins/database.py#L144-L148).

It's easy to fix, though.  `provision_mysql()` calls three methods, and only `_create_database` would need to be changed:
* [`_create_database`](https://github.com/open-craft/opencraft/blob/f94660e9e267f85af61ceb50c6600003998b6276/instance/models/mixins/database.py#L69-L74) - could be modified to be [`CREATE DATABASE <database> IF NOT EXISTS`](http://dev.mysql.com/doc/refman/5.7/en/create-database.html). 
* [`create_user`](https://github.com/open-craft/opencraft/blob/f94660e9e267f85af61ceb50c6600003998b6276/instance/models/mixins/database.py#L77-L86) - creates the given user only if it does not already exist.
* [`_grant_privileges`](https://github.com/open-craft/opencraft/blob/f94660e9e267f85af61ceb50c6600003998b6276/instance/models/mixins/database.py#L89-L97) - GRANT is a no-op if the privileges are already granted.

66c50f0 makes this change, but it raises the question: do we need the `mysql_provisioned` flag anymore?

Comment (mtyaka) [inline instance/models/mixins/openedx_database.py]: Nice improvement, will make adding new databases in the future easier.

> do we need the mysql_provisioned flag anymore?

I don't think we do. The methods that deprovision databases already seem to be designed so that they won't fail if the database or user does not exist, so `mysql_provisoined` flag doesn't make much sense anymore.

Comment (pomegranited) [inline instance/models/mixins/openedx_database.py]: Also looks like we can remove `mongo_provisioned`, because if [`mongo.add_user`](https://github.com/open-craft/opencraft/blob/f94660e9e267f85af61ceb50c6600003998b6276/instance/models/mixins/database.py#L223) is called on an existing user, all it does is potentially [update the password](http://api.mongodb.com/python/current/api/pymongo/database.html#pymongo.database.Database.add_user).

And [`swift_provisioned`](https://github.com/open-craft/opencraft/blob/077e1a71dcfc574581b4c11581794d9b8436d45b/instance/models/mixins/storage.py#L59) too!  If the container already exists, it might only [update its metadata](http://docs.openstack.org/developer/python-swiftclient/swiftclient.html#swiftclient.client.post_container).

Does that sound ok to you @mtyaka ?

**Comment by pomegranited** [comment]
@mtyaka I've updated the PR instructions to ensure the 2nd instance provisions correctly.

**Comment by pomegranited** [comment]
@mtyaka my [previous test failure](https://circleci.com/gh/open-craft/opencraft/1878) was due to a strange error:

```
psycopg2.OperationalError: database "test_opencraft" is being accessed by other users
DETAIL:  There is 1 other session using the database.
```

I've never seen that before.. There were no other builds running with mine.  Do you think it could be a side effect of this change?  I'm [re-running now to see if it goes away](https://circleci.com/gh/open-craft/opencraft/1879), but would hate to introduce any flakiness.

**Comment by mtyaka** [comment]
:+1: The instance that I was trying to spawn failed due some kind of timeout, but it was past the `programs` role, so that shouldn't block this PR from getting merged.

- [x] I tested this: Spawned a new instance using settings from the PR description. Without this patch, ansible failed on the programs step; with this patch it performed the programs steps successfully. Spawned another app server to ensure reprovisioning mysql, mongo, swift works on existing instance.
- [x] I read through the code

**Comment by bradenmacdonald** [comment]
@pomegranited Nit: Can we silence all these things which kind of look like errors in the log? Or catch this exception and make it nicer like `Checking database... already exists, good.`

```
/home/ubuntu/opencraft/instance/models/mixins/database.py:74: Warning: (1007, "Can't create database 'instance0de2cc7f_test_example_com_ecommerce'; database exists")
  cursor.execute('CREATE DATABASE IF NOT EXISTS `{database}` DEFAULT CHARACTER SET utf8'.format(database=database))
/home/ubuntu/opencraft/instance/models/mixins/database.py:74: Warning: (1007, "Can't create database 'instance0de2cc7f_test_example_com_dashboard'; database exists")
  cursor.execute('CREATE DATABASE IF NOT EXISTS `{database}` DEFAULT CHARACTER SET utf8'.format(database=database))

...
18/Jan/2017 04:45:19 |     INFO | process=26320 | swiftclient               | REQ: curl -i https://storage.gra1.cloud.ovh.net/v1/AUTH_2b26f56d1c394ca88ee2b412720d73de/1ec9cc4c_integration_plebia_net?format=json -X GET -H "X-Auth-Token: 660fc0aca05342c9..." -H "Accept-Encoding: gzip"
18/Jan/2017 04:45:19 |     INFO | process=26320 | swiftclient               | RESP STATUS: 404 Not Found
18/Jan/2017 04:45:19 |     INFO | process=26320 | swiftclient               | RESP HEADERS: {'Date': 'Wed, 18 Jan 2017 04:45:19 GMT', 'X-Trans-Id': 'tx722d9e04a92a437d9dfd2-00587ef2de', 'Content-Length': '70', 'Content-Type': 'text/html; charset=UTF-8'}
18/Jan/2017 04:45:19 |     INFO | process=26320 | swiftclient               | RESP BODY: b'<html><h1>Not Found</h1><p>The resource could not be found.</p></html>'
18/Jan/2017 04:45:19 |    ERROR | process=26320 | swiftclient               | Container GET failed: https://storage.gra1.cloud.ovh.net/v1/AUTH_2b26f56d1c394ca88ee2b412720d73de/1ec9cc4c_integration_plebia_net?format=json 404 Not Found  [first 60 chars of response] b'<html><h1>Not Found</h1><p>The resource could not be found.<'
Traceback (most recent call last):
  File "/home/ubuntu/virtualenvs/venv-3.5.1/lib/python3.5/site-packages/swiftclient/client.py", line 1647, in _retry
    service_token=self.service_token, **kwargs)
  File "/home/ubuntu/virtualenvs/venv-3.5.1/lib/python3.5/site-packages/swiftclient/client.py", line 917, in get_container
    raise ClientException.from_response(resp, 'Container GET failed', body)
swiftclient.exceptions.ClientException: Container GET failed: https://storage.gra1.cloud.ovh.net/v1/AUTH_2b26f56d1c394ca88ee2b412720d73de/1ec9cc4c_integration_plebia_net?format=json 404 Not Found  [first 60 chars of response] b'<html><h1>Not Found</h1><p>The resource could not be found.<'
```

Also this is an optional nit, but it would be nice to combine these three migrations into one before merging this.

File: instance/tests/models/test_openedx_database_mixins.py [hunk 2]
```diff
@@ -470,6 +479,9 @@ def test_ansible_settings_mongo(self):
         self.assertIn('FORUM_MONGO_PORT: {0}'.format(MONGODB_SERVER_DEFAULT_PORT), ansible_vars)
         self.assertIn('FORUM_MONGO_DATABASE: {0}'.format(self.instance.forum_database_name), ansible_vars)
 
+        # Unset self.instance to avoid tearDown failures trying to deprovisioning databases that were not created
+        self.instance = None
```
Comment (bradenmacdonald) [inline instance/tests/models/test_openedx_database_mixins.py]: Why are these lines needed? I thought Matjaz said that "The methods that deprovision databases already seem to be designed so that they won't fail if the database or user does not exist"

If I go into the shell and I go:
```python
instance = production_instance_factory(name="Production instance", sub_domain="production")
instance.delete()
```
Does that still work without error (no actual provisioning or deprovisioning happens)?

**Comment by pomegranited** [comment]
@mtyaka  Apologies for how long this PR has taken me to do..  I was still getting those weird errors about being unable to drop the `test_opencraft` database, even after removing my test changes.  So I decided to simplify the change, to see if that fixes it.  *fingers crossed*

You'll need to revert the migrations on your devstack to re-test this, I'm afraid.

```
make manage migrate instance 0067_remove_blank_name
make migrate
```

* Left the `*_provisioned` flags in place.  I'm not checking them on provisioning, so added databases, tables, and storage can be added after initial appserver creation, but I'm still checking them on deprovisioning.
* Using `IF NOT EXISTS` during MySQL database creation, but am silencing warnings that occur during that call. (CC @bradenmacdonald).
* I haven't been able to silence the red herring `ERROR` logs generated by the swift library though.  They've been happening all along, because the swift library checks for the existence of the container before trying to create it.

I've also rebased on the latest master, to assist with the merge.  If you're running the tests in your devstack, note the changes to your .env that are required since https://github.com/open-craft/opencraft/pull/152 merged, and the temporary change which may be required to your .env.test: https://github.com/open-craft/opencraft/pull/152#issuecomment-274225119.

**Comment by mtyaka** [comment]
👍 I re-tested this and the changes make sense. You'll have to rebase again, though.

**Comment by pomegranited** [comment]
Thanks, @mtyaka , but oh no!  I know that if I rebase, I'll never get my tests to pass again..  ref https://github.com/open-craft/opencraft/pull/169 😢
